MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,247 @@
+Loose Object Cleanup Action
+---------------------------
+
+The loose object action can save users who have many loose object downloads.
+These loose objects are populated by several user activities: running
+`git fetch` will get loose commits and trees for the ref tips that are not
+covered by prefetch packs, hydrating a file will trigger a loose blob, and many
+other. Older enlistments have many loose commits and trees due to the behavior
+of `git fetch` before the `--no-show-forced-updates` option.
+
+### Delete loose objects that appear in a pack-file
+
+By running `git prune-packed -q` with the environment variable
+`GIT_OBJECT_DIRECTORY` set to the object cache, Git will delete the loose
+objects that exist in a pack-file. This will use the multi-pack-index to test
+containment, so is only bound by the file-deletion operations.
+
+### Pack loose objects that do not appear in a pack-file
+
+We can use `git pack-objects -q <object-dir>/pack/fromloose-<timestamp>` to create a
+pack-file filled with a list of loose objects. We can pass the list of loose
+objects to this command through stdin. We can choose to batch this operation by
+limiting the number of objects we pack at a time, if we want. We may also want
+to skip this step if the number of loose objects is low (&lt; 1,000? &lt;
+10,000?)
+
+By listing a timestamp in the pack-file name, we can store the "most-recent" 
+time this job was run.
+
+*Note:* While we are thinking about packing loose objects, we can consider
+_deleting_ some loose blobs. It is easier to delete loose blobs than packed
+blobs. If we want to delete "stale" blobs, then we should add a step to this
+job.
+
+Prefetch Pack Cleanup Action
+----------------------------
+
+We store a list of prefetch packs as `<object-dir>/pack/prefetch-<timestamp>-<sha>.[pack|idx]`.
+These are downloaded at a regular cadence by the post-fetch job, and could have
+a new pack-file for every hour. In addition to these prefetch packs, some daily
+prefetch packs will appear and occasionally a very large prefetch pack covering
+everything more than 30 days old. This large prefetch pack appears on initial
+clone and sometimes again if a user does not prefetch for 30 days (this occurs 
+less often now that background prefetch exists).
+
+### Delete pack-files with no objects referenced by the multi-pack-index
+
+The multi-pack-index stores a single copy of every packed object. When an object
+appears in multiple pack-files covered by the multi-pack-index, we select the
+pack-file with the most-recent _modified time_ as reported by the filesystem.
+For example, if someone downloads a second large prefetch pack when prefetching
+for the first time in 31 days, then that large prefetch pack is referenced and
+all older prefetch packs are "stale" and not referenced by the multi-pack-index.
+
+We can extend the `git multi-pack-index` builtin to delete pack-files that have
+no objects referenced by the multi-pack-index. This should happen at the same

```

Comment:
If there is a pack file with only a few objects referenced by the multi-pack-index would it be possible (and if so, should we) give it priority when deciding what to re-pack?

Reply:
That's an interesting idea! If we go this route, we want to specifically target the objects that are uniquely in those pack-files. This may be a better way to go (instead of repacking entire pack-files).

Here's how I see the approach you suggest working:

1. We create a `git multi-pack-index repack` verb (or a special mode for `git repack` that is multi-pack-index aware).
2. This Git process inspects the counts of packs marked by the multi-pack-index, and finds certain conditions for a pack being "stale" (say, modified time is old enough or there are at most X referenced objects).
3. The git process calls `git pack-objects`, passing the list of _objects_ (not the list of pack-files) and creates a pack.
4. The git process then updates the multi-pack-index with pack-level metadata saying the new pack-files are "repacked" pack-files and the old pack-files are "expired".
5. We keep the expired pack-files in the multi-pack-index until we repack again.

This is something I can really dig into and see what is possible. I'll want to make an RFC to the mailing list, as this seems like a big reason to have the multi-pack-index. In fact, [making `repack` use the `multi-pack-index` in a smart way is part of the design document's "Future Work" section](https://github.com/git/git/blob/master/Documentation/technical/multi-pack-index.txt#L86-L93).

(The good news is that I can let that idea simmer for a while as we work on the refactor and loose-object improvements.)